src/LanguageServer/Protocol/Handler/Extensions/ExtensionDocumentMessageHandler.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Remote;
+using Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Extensions;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(ExtensionDocumentMessageHandler)), Shared]
+[Method(MethodName)]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class ExtensionDocumentMessageHandler()
+    : ILspServiceDocumentRequestHandler<ExtensionDocumentMessageParams, ExtensionMessageResponse>
+{
+    private const string MethodName = "roslyn/extensionDocumentMessage";
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(ExtensionDocumentMessageParams request)
+    {
+        return request.TextDocument;
+    }
+
+    public async Task<ExtensionMessageResponse> HandleRequestAsync(ExtensionDocumentMessageParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Document);
+        Contract.ThrowIfNull(context.Solution);
+
+        var solution = context.Solution;
+        var project = context.Document.Project;
+        var client = await RemoteHostClient.TryGetClientAsync(project, cancellationToken).ConfigureAwait(false);
+
+        if (client is not null)
+        {
+            var response = await client.TryInvokeAsync<IRemoteExtensionMessageHandlerService, string>(
+                project,
+                (service, solutionInfo, cancellationToken) => service.HandleExtensionDocumentMessageAsync(
+                    solutionInfo,
+                    request.MessageName,
+                    request.Message,
+                    context.Document.Id,
+                    cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!response.HasValue)
+            {
+                throw new InvalidOperationException("The remote message handler didn't return any value.");
+            }
+
+            return new ExtensionMessageResponse(response.Value);
+        }
+        else
+        {
+            Contract.ThrowIfNull(context.Workspace);
```

Comment:
i would not do this.  effectively, we *must* get this loaded in OOP, as we have no local way to do ALCs at all.  so if we can't get the remote client, we just error out.

Reply:
We shouldn't error out - we have no OOP in vscode, but this works fine since we run on .net9

Reply:
gotcha.  ok.  we need to change this to just call the service. the service attempts the remote call, falling back to in-proc if it can't.

Reply:
I don't understand what I have to change here